bin/pytorch_inference/Main.cc

```diff
@@ -4,104 +4,117 @@
  * you may not use this file except in compliance with the Elastic License.
  */
 
+#include <api/CIoManager.h>
 #include <core/CBlockingCallCancellingTimer.h>
 #include <core/CLogger.h>
 #include <core/CProcessPriority.h>
 #include <core/CRapidJsonLineWriter.h>
-#include <core/CoreTypes.h>
-
-#include <ver/CBuildInfo.h>
-
-#include <api/CIoManager.h>
-
 #include <seccomp/CSystemCallFilter.h>
+#include <ver/CBuildInfo.h>
 
 #include "CBufferedIStreamAdapter.h"
 #include "CCmdLineParser.h"
+#include "CCommandParser.h"
 
 #include <rapidjson/ostreamwrapper.h>
 #include <torch/script.h>
 
-#include <boost/optional.hpp>
-
 #include <memory>
 #include <string>
 
-// For ntohl
-#ifdef Windows
-#include <WinSock2.h>
-#else
-#include <netinet/in.h>
-#endif
+static const std::string INFERENCE{"inference"};
+static const std::string ERROR{"error"};
 
-using TFloatVec = std::vector<float>;
+torch::Tensor infer(torch::jit::script::Module& module,
+                    ml::torch::CCommandParser::SRequest& request) {
 
-torch::Tensor infer(torch::jit::script::Module& module, TFloatVec& data) {
     torch::Tensor tokensTensor =
-        torch::from_blob(data.data(), {1, static_cast<std::int64_t>(data.size())})
+        torch::from_blob(static_cast<void*>(request.s_Tokens.data()),
+                         {1, static_cast<std::int64_t>(request.s_Tokens.size())},
+                         at::dtype(torch::kInt32))
             .to(torch::kInt64);
+
     std::vector<torch::jit::IValue> inputs;
     inputs.push_back(tokensTensor);
-    inputs.push_back(torch::ones({1, static_cast<std::int64_t>(data.size())})); // attention mask
-    inputs.push_back(
-        torch::zeros({1, static_cast<std::int64_t>(data.size())}).to(torch::kInt64)); // token type ids
-    inputs.push_back(torch::arange(static_cast<std::int64_t>(data.size())).to(torch::kInt64)); // position ids
+
+    for (auto args : request.s_SecondaryArguments) {
+        torch::Tensor tensor =
+            torch::from_blob(static_cast<void*>(args.data()),
+                             {1, static_cast<std::int64_t>(args.size())},
+                             at::dtype(torch::kInt32))
+                .to(torch::kInt64);
+
+        inputs.push_back(tensor);
```

Comment:
This is copying the tensors at the moment.  It may be possible to emplace:

```suggestion
        inputs.emplace_back(
            torch::from_blob(static_cast<void*>(args.data()),
                             {1, static_cast<std::int64_t>(args.size())},
                             at::dtype(torch::kInt32))
                .to(torch::kInt64));
```

Or if that doesn't work for some reason, you could at least move when adding, i.e. `inputs.push_back(std::move(tensor));`.

Reply:
`torch::tensor` is a wrapper around `at::tensor` and internally has a smart pointer to its data. Passing by value/copying is the way to use it by design

Some discussion here https://discuss.pytorch.org/t/tensor-move-semantics-in-c-frontend/77901/5